### DISCUSSION TITLE
QueryClient DefaultOptions error type

### DISCUSSION THREAD

**Discussion by akinzalowevidation** [TanStack/query#2757] [Ideas] [answered]
Hi, 

I am wondering why it is not possible to define the error type of the `DefaultOptions` within a `QueryClient`?

I am using it in this way:

```
const handleError: (error: DefaultGenericError) => {
  ...
};

const queryClient = new QueryClient({
  defaultOptions: {
     queries: {
       onError: handleError,
     },
    mutations: {
      onError: handleError,
     },
   }}
});
```

Would `new QueryClient<DefaultGenericError>(...)` (which is passed to `DefaultOptions`) be reasonable here as a suggestion?

**Comment by TkDodo** [comment, marked as answer]
`error` defaults to `unknown` because everything can be `thrown` in javascript. If you're certain that error is _always_ of a specific type, you can do an explicit type assertion:

```
const handleError = (error: unknown) => {
  const genericError = error as DefaultGenericError
};
```

the safer way would be to let typescript perform type narrowing by doing runtime checks:

```
const handleError = (error: unknown) => {
  if (error instanceOf Error) {
    // do something with error.message
  }
};
```

I've also written about that here: https://tkdodo.eu/blog/react-query-and-type-script#what-about-error
